### ISSUE TITLE
There is no ability to use MachineOptions (WithYear) with nontransparent/octetcounting parsers

### ISSUE THREAD

**Issue by dehaansa** [leodido/go-syslog#21]
This library is used in grafana/loki and then that code is used within grafana/alloy as well, and a user encountered an issue with the default 0'd out year with a RFC3164 message https://github.com/grafana/alloy/issues/2287 . I went to go update the loki code to use the WithYear/CurrentYear option but [that code](https://github.com/grafana/loki/blob/main/clients/pkg/promtail/targets/syslog/syslogparser/syslogparser.go) uses the nontransparent & octetcounting parsers, rather than the rfc3164 & rfc5424 parsers.

I would propose updating the parsers to support these options. I think an ideal world would unify the With<Option> to no longer have separate ParserOption & MachineOption function types, but at least support for WithYear in the other parsers would unblock these changes.

**Comment by leodido** [comment]
Hi @dehaansa, let me see if I understood correctly.

The problem `grafana/alloy` is facing is not being able to parse syslog messages (using either octet-counting or non-transparent framing)  using a `WithYear(CurrentYear{})` option strategy.

You folks need something like this but for non-transparent & octet counting parsers... Correct?

https://github.com/leodido/go-syslog/blob/46afcecb86058ec9598480d6140f404266a02868/rfc3164/example_test.go#L41

**Comment by dehaansa** [comment]
> Hi [@dehaansa](https://github.com/dehaansa), let me see if I understood correctly.
> 
> The problem `grafana/alloy` is facing is not being able to parse syslog messages (using either octet-counting or non-transparent framing) using a `WithYear(CurrentYear{})` option strategy.
> 
> You folks need something like this but for non-transparent & octet counting parsers... Correct?
> 
> [go-syslog/rfc3164/example_test.go](https://github.com/leodido/go-syslog/blob/46afcecb86058ec9598480d6140f404266a02868/rfc3164/example_test.go#L41)
> 
> Line 41 in [46afcec](/leodido/go-syslog/commit/46afcecb86058ec9598480d6140f404266a02868)
> 
>  p := NewParser(WithYear(CurrentYear{}))

Yup, that's exactly right. When I wrap up some other tasks I intend to come back and implement it, if no one else takes a look at it first. Thanks.

EDIT: Just saw that you opened a PR! Looks right, will validate next week.

**Comment by lowjoel** [comment]
Hello there! I did an equivalent change for influxdata/telegraf#17156 and would love for the changes to be made available. Have the Grafana folk replied?
